CLI cleanups

jonatack opened the pull request:
These have been accumulating over the past few years.

Each is described in its commit message.

jonatack committed:
cli, refactor: simplify DetailsRequested()

The bounds check is no longer needed after the merge of PR 21192.

jonatack committed:
cli, refactor: deduplicate NetworkStringToId()

jonatack committed:
cli, refactor: simplify public-only classes with structs

and run clang-format on it

hodlinator commented on src/bitcoin-cli.cpp:
```diff
@@ -130,14 +130,11 @@ static void libevent_log_cb(int severity, const char *msg)
 // when to wait if -rpcwait is given.
 //
-class CConnectionFailed : public std::runtime_error
+struct CConnectionFailed : std::runtime_error
```
Weird to change inheritance from `class -> class` to `struct -> class`.

jonatack replied:
Thanks for having a look! Replied below in https://github.com/bitcoin/bitcoin/pull/31887#discussion_r1966566982.

hodlinator replied:
I consider mixing `class`/`struct` in the inheritance hierarchy to be a separate issue.

jonatack replied:
See https://stackoverflow.com/questions/3574040/c-can-a-struct-inherit-from-a-class in addition to my response above.

hodlinator replied:
My objection does not come from a knowledge gap in what C++ can and cannot do. It is about sticking to certain conventions.

As I was checking to see if the Core Guidelines had influenced me in the matter I only found weak counter evidence:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c122-use-abstract-classes-as-interfaces-when-complete-separation-of-interface-and-implementation-is-needed
In the example code `class D1` inherits from `struct Device`, full of virtual methods.

jonatack replied:
Yes, here it seems cleaner, fitting the use case and requiring less code to do the same thing. I appreciate your indulgence here.

hodlinator commented on src/bitcoin-cli.cpp:
```diff
@@ -267,21 +264,19 @@ static int8_t NetworkStringToId(const std::string& str)
  * as well as converting back to a JSON object that can be shown as result.
  */
-class BaseRequestHandler
+struct BaseRequestHandler
```
Makes more sense to use `class`es for these as they use `virtual` functions. Not worth changing just to be able to remove `public` IMO.

jonatack replied:
If I'm understanding the feedback, it comes from the idea of C structs being for POD. My understanding is that in C++, classes and structs are identical except for their default inheritance and member access levels, and for the cases here, structs seem to be the appropriate choice that allows simpler code. See also https://stackoverflow.com/a/46339933. I don't mind dropping this change if there is a compelling reason that I'm not aware of, though.

hodlinator replied:
> If I'm understanding the feedback, it comes from the idea of C structs being for POD. My understanding is that in C++, classes and structs are identical except for their default inheritance and member access levels, and for the cases here, structs seem to be the appropriate choice that allows simpler code.

I agree that C++ classes and structs are essentially the same, no confusion there. (C# disallows inheritance for `struct` but not `class`, Rust doesn't allow inheritance, period. I know we're in C++, but still aware of the tendencies).

What contributed to giving me a sense that this project attempts to use `struct`/`class` as a shorthand for distinct things (like POD/vs non-POD), beyond the code that passed me by so far, was dev-notes.md:

https://github.com/bitcoin/bitcoin/blob/e486597f9a57903600656fb5106858941885852f/doc/developer-notes.md?plain=1#L70-L71

I want to maintain a fairly equal bar for review and 3252f3bab02605129a568112f650ef3cb29703b7 (the topic of this thread) stands out to me as having been shot down instantly if done by a relative newcomer. I've benefited from [your review guide](https://jonatack.github.io/articles/on-reviewing-and-helping-those-who-do-it) and am sure you've provided and provide much value to the project.

I really hope I can build a reputation for other things than nits. :)

---

If you insist on keeping structs, maybe hoist up the braces to conform with the dev-notes? Also, you have inconsistent spacing:
`struct AddrinfoRequestHandler : BaseRequestHandler` vs
`struct GetinfoRequestHandler: BaseRequestHandler`

jonatack replied:
> hoist up the braces to conform with the dev-notes? Also...inconsistent spacing

>  it would be good to use recommended style (can just run `clang-format`)

Good idea, done.

hodlinator reviewed (commented):
Code reviewed 3252f3bab02605129a568112f650ef3cb29703b7

ACK 9184ec9e8ea6efd4d0d3625a06f223156f940243

Advise dropping 3rd commit, the others seem good.

ryanofsky commented on src/bitcoin-cli.cpp:
re: https://github.com/bitcoin/bitcoin/pull/31887#discussion_r1959923000

Mostly agree with this suggestion. It would be nice to remove this `IsArgSet` call (in a separate commit or PR), because like [many](https://github.com/bitcoin/bitcoin/pull/30529) [other](https://github.com/bitcoin/bitcoin/pull/31212) `IsArgSet` calls, this one introduces a bug and doesn't handle negation properly. A common sense interpretation of `-nogetinfo` would be to run the getinfo command, not to run it.

Only part of suggestion above I'd disagree with is the part about also adding the `DISALLOW_NEGATION` flag. I think the `DISALLOW_NEGATION` flag is useful for non-boolean options accepting string/int/path/list values where not having a value isn't supported or doesn't make sense (e.g. `-nodatadir`). For other options, it is better to stay consistent and support negation so code does not fill up with special cases and users do not have to guess which options support negation and which will trigger errors for no particular reason.

yancyribbens replied:
>  It would be nice to remove this IsArgSet call (in a separate commit or PR)

Looks like that was followed up on already? https://github.com/bitcoin/bitcoin/pull/31896

jonatack replied:
> It seems odd to treat -getinfo different than -netinfo (and everything else). It would be good for consistency to treat them the same

Thanks, done.

ryanofsky reviewed (approved):
Code review ACK 3252f3bab02605129a568112f650ef3cb29703b7. All changes look good. No real opinion on the third commit, seems fine to keep or drop

yancyribbens commented on src/bitcoin-cli.cpp:
```diff
@@ -404,7 +404,7 @@ class NetinfoRequestHandler : public BaseRequestHandler
         return UNKNOWN_NETWORK;
     }
     uint8_t m_details_level{0}; //!< Optional user-supplied arg to set dashboard details level
-    bool DetailsRequested() const { return m_details_level > 0 && m_details_level < 5; }
```
> The upper bound check is no longer needed after the merge of PR 21192.

Nit, looks like the bound check is for upper and lower bounds (above zero and bellow 5).

jonatack replied:
Thank you for taking a look! `m_details_level` is capped at a maximum value of 4 since https://github.com/bitcoin/bitcoin/pull/21192/files; that line is currently `m_details_level = std::min(n, NETINFO_MAX_LEVEL);`

yancyribbens replied:
I think you misunderstood my comment (or I am not understanding yours).  min provides both an upper an lower bound here, so it's just bounded.  In the commit you say "upper bound" but I'm saying it should just read "bound".

jonatack replied:
Oh ok, updated the commit message per s/upper bound/bound/ as you suggested.

yancyribbens commented:
Concept Ack https://github.com/bitcoin/bitcoin/commit/3252f3bab02605129a568112f650ef3cb29703b7

jonatack committed:
cli, bugfix: for -getinfo, replace IsArgSet() with GetBoolArg()

for consistency with the other CLI commands (-netinfo, -addrinfo, -generate).

This can be considered a bugfix because IsArgSet() returns whether an arg has
been set even if it has been negated. After this change, we no longer treat
-nogetinfo and -getinfo=0 the same as -getinfo and -getinfo=1, and instead as if
-getinfo was not specified.

Co-authored-by: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>
Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

pablomartin4btc reviewed (commented):
cr ACK ccc34a26170ef8c63a6f211eda4464e954873e89

Regarding the 3rd [commit](3252f3bab02605129a568112f650ef3cb29703b7):
- for the exceptions case, perhaps this can be done in a separate PR as I can see other places where we are using `class` outside `CLI` if we want to be consistent (and if we think we won't require encapsulation for them).
- for the `*RequestHandler`, are we sure the internal representation of these objects and what they expose won't change at any time in the future? It looks like it should be ok, but I'm not too sure if at some point we would need to.

ryanofsky reviewed (approved):
Code review ACK ccc34a26170ef8c63a6f211eda4464e954873e89. Since last review just added a commit replacing IsArgSet call. I left some suggestions that could be followed up, but they are not very important.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -1252,7 +1252,7 @@ static int CommandLineRPC(int argc, char *argv[])
         std::unique_ptr<BaseRequestHandler> rh;
         std::string method;
-        if (gArgs.IsArgSet("-getinfo")) {
+        if (gArgs.GetBoolArg("-getinfo", false)) {
```
Thank you for clarifying the change. Done.

l0rinc replied:
This is a surprising change, the PR states this is a refactor - I would prefer doing this separately

jonatack replied:
It was added at the request of multiple reviewers and doesn't seem worth splitting out to another PR for that one line and requiring an additional merge script run. Instead, I've dropped the word "refactor" from the PR title.

jonatack commented:
Thank you for reviewing, @pablomartin4btc. I prefer to keep this pull scoped only to this file and not increase the scope, as these changes are local to it and it is a file I know reasonably well (the RequestHandler representation hasn't changed over that time period).

jonatack commented:
Thanks for the reviews, everyone. Ran clang-format on the third commit and adjusted a few commit messages per the feedback. Should be trivial to re-ack (thanks!)

pablomartin4btc reviewed (commented):
re-ACK d423fd9ec83ae323ac29bdc1b677ed8260bd59c4

hodlinator reviewed (approved):
Code review ACK d423fd9ec83ae323ac29bdc1b677ed8260bd59c4

Good to see `-getinfo` negation behavior fixed. I encountered it when exploring args in https://github.com/bitcoin/bitcoin/pull/16545#pullrequestreview-2316714013 (in the "Expand/collapse lengthy exploration"-section).

Still think the 3rd commit is out of character compared to my brief experience with this project as it is purely stylistic, but at least the formatting was [fixed](https://github.com/bitcoin/bitcoin/pull/31887#discussion_r1992220870) since my previous review.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -396,15 +385,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
-    }
     uint8_t m_details_level{0}; //!< Optional user-supplied arg to set dashboard details level
-    bool DetailsRequested() const { return m_details_level > 0 && m_details_level < 5; }
+    bool DetailsRequested() const { return m_details_level; }
```
I omit the `#` in cases like this, where it isn't really important to link them here, so that GitHub does not add a notification in that PR every time this branch is (re)pushed.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -259,30 +255,25 @@ static void http_error_cb(enum evhttp_request_error err, void *ctx)
 }
 
-/** Class that handles the conversion from a command-line to a JSON-RPC request,
+static int8_t NetworkStringToId(const std::string& str)
```
Your review comment provides that context (thanks). I wouldn't usually add that to the commit message for the same reason as https://github.com/bitcoin/bitcoin/pull/31887#discussion_r1992475317. In this case, however, where the PR discussion is locked, maybe it wouldn't add notifications there. Not sure.

l0rinc commented on src/bitcoin-cli.cpp:
```diff
@@ -404,7 +404,7 @@ class NetinfoRequestHandler : public BaseRequestHandler
     }
     uint8_t m_details_level{0}; //!< Optional user-supplied arg to set dashboard details level
-    bool DetailsRequested() const { return m_details_level > 0 && m_details_level < 5; }
+    bool DetailsRequested() const { return m_details_level; }
```
This is better than before, but having a `NETINFO_MAX_LEVEL` and a checking the bounded values in methods and having `DetailsRequested`, but also checking it manually in https://github.com/bitcoin/bitcoin/blob/master/src/bitcoin-cli.cpp#L499-L501 where we're checking the raw `n` value seems quite hacky.

Seems the code is begging for a bounded type - if we're cleaning it up, we might as well try an enum here:

```C++
enum class NetinfoDetailLevel : uint8_t {
    BASIC = 0,
    PEERS,
    PEERS_WITH_ADDR,
    PEERS_WITH_VER,
    FULL
};
```

jonatack replied:
> where we're checking the raw n value

That is to check user input.

I think the abstraction level here is fine, but happy to review if you propose a further abstraction in a follow-up.

l0rinc commented on src/bitcoin-cli.cpp:
```diff
@@ -259,6 +259,14 @@ static void http_error_cb(enum evhttp_request_error err, void *ctx)
     reply->error = err;
 }
 
+static int8_t NetworkStringToId(const std::string& str)
```
from the usages it seems we can modernize this in a few more ways:
* `UNKNOWN_NETWORK` seems to serve as a value-missing proxy - we could use an optional instead
* `const std::string&` could be modernized to `static std::optional<int8_t> NetworkStringToId(std::string_view network)`


This would allow us to change the usages to something like:
```C++
if (auto network_id{NetworkStringToId(network_name)}) {
    ++counts.at(*network_id);
}
```

jonatack replied:
Yes, IIRC `std::optional` wasn't available when this was added, as the codebase was using C++11 at that time. I think this would be more of a modernization rewrite than a cleanup, but happy to review if you want to propose it.

l0rinc commented on src/bitcoin-cli.cpp:
```diff
@@ -259,6 +259,14 @@ static void http_error_cb(enum evhttp_request_error err, void *ctx)
 
+static int8_t NetworkStringToId(const std::string& str)
+{
+    for (size_t i = 0; i < NETWORKS.size(); ++i) {
```
we're returning an `int8_t`, we might as well iterate over one
```suggestion
    for (int8_t i{0}; i < NETWORKS.size(); ++i) {
```

jonatack replied:
Not sure but I think this would cause a conversion from size_t to int8_t on each iteration, rather than once for `return i`. Doesn't seem worth it.

l0rinc replied:
the problem is likely that the method return `int8_t` for an index - unnecessary, should likely be `size_t` or `int` - not important

l0rinc commented on src/bitcoin-cli.cpp:
```diff
@@ -267,21 +263,17 @@ static int8_t NetworkStringToId(const std::string& str)
-class AddrinfoRequestHandler : public BaseRequestHandler
-{
-public:
+struct AddrinfoRequestHandler : BaseRequestHandler {
```
simple structs are fine, but I'm not a fan of struct inheritance, I like the previous version more  in the non-trivial cases

jonatack replied:
If I'm understanding the comment ("simple structs are fine"), it comes from the idea of structs being for POD, say, in C. In C++, classes and structs are identical except for their default inheritance and member access levels, and for the use cases here, a struct appears to be the more appropriate choice that allows simpler, cleaner code (see also [stackoverflow.com/a/46339933](https://stackoverflow.com/a/46339933). Struct inheritance is also fine (see discussions like [stackoverflow.com/questions/3574040/c-can-a-struct-inherit-from-a-class](https://stackoverflow.com/questions/3574040/c-can-a-struct-inherit-from-a-class). If there is a technical issue beyond style nits, happy to re-evaluate.

l0rinc reviewed (changes requested):
I like these cleanups, ans we should go a step further and modernize the code more when touching them.
I'm also not a fan of the 3rd commit for the inheritance cases and would prefer seeing the last commit in a separate PR, doesn't seem like a "refactor:"

jonatack commented:
> I like these cleanups, ans we should go a step further and modernize the code more when touching them. I'm also not a fan of the 3rd commit for the inheritance cases and would prefer seeing the last commit in a separate PR, doesn't seem like a "refactor:"

The last commit was added at the request of multiple reviewers, but I've removed "refactor" from the pull title. Addressed the other feedback individually and happy to look at any modernization follow-ups you'd like to propose.

l0rinc reviewed (approved):
I would prefer doing the modernizations here to get rid of more technical debt, but this is already better than before so:

ACK d423fd9ec83ae323ac29bdc1b677ed8260bd59c4

ryanofsky reviewed (approved):
Code review ACK d423fd9ec83ae323ac29bdc1b677ed8260bd59c4, just running clang-format and updating commit messages since last review
